### PR TITLE
feat(console): o opens the selected resource in the AWS console, O copies the URL

### LINKED ISSUE #477: feat: "o" opens the highlighted resource in the AWS console (list + detail)
## Summary

a9s is read-only by design. When you find a problem in a list (a red row, a failed instance, an expiring cert), the next step is always the same: alt-tab to a browser, click through the AWS console's navigation maze, pick the right region, and search for the resource you were already looking at. That round-trip is the single most annoying seam in the workflow.

**Feature:** pressing `o` on a highlighted resource — in a resource list or in the detail view — opens the default browser directly on that resource's AWS console page, with the correct region, service console, and partition. Companion binding: `O` copies the same console URL to the clipboard (for pasting into Slack/tickets, or when the terminal is on a remote host).

Both `o` and `O` are currently unbound anywhere (TUI keymap, web keyMap, catalog child-view triggers, help overlay — verified exhaustively).

## UX

| Context | `o` | `O` |
|---|---|---|
| Resource list (top-level type) | open console page of highlighted row | copy console URL |
| Detail view | open console page of the detailed resource | copy console URL |
| Detail view, related panel focused | open console page of the selected related row | copy its URL |
| Child-view list (log streams, executions, …) | open page if the child type has a URL mapping, else flash "no console page" | same |
| Menu / help / costs / selector screens | no-op (key not applicable — gated by screen kind) | no-op |
| Filter `/`, command `:`, search input modes | types the letter `o` into the input (existing input-mode guards already run first) | same |
| Demo mode (`--demo`) | flash "demo mode — no real console resource", do not open browser | copy still works (URL is well-formed, account is synthetic) |

Feedback: on success, `Flash` toast "opened in AWS console" (TUI) — mirroring the existing `c` copy toast. On failure (no URL mapping, opener failed), `Flash` error. Never crash, never block the UI.

## Architecture sketch

Follows the house pattern end to end; closest precedent is the `c` copy key (per-type `CopyField` catalog config + renderer-only side effect in a `tea.Cmd`).

1. **URL building is pure and lives in core** — new small package (e.g. `core/consolelink`): `Build(typeDef, resource, region, accountID) (url string, ok bool)`. Unit-testable with zero AWS calls.
2. **Per-type deterministic config in the catalog** — new field on `catalog.ResourceTypeDef` (Identity section, next to `CostExplorerServiceName`, `core/catalog/types.go:39-44`): either a template string with placeholders (`{region}`, `{id}`, `{name}`, `{account}`, `{field:xyz}` — same mini-DSL spirit as `CloudTrailKey`) or a `func(r domain.Resource, region, account string) string` for the few types that need branching (see edge cases). One line per type in each `core/aws/catalog_<category>.go` literal. Per-type config, not reflection/heuristics.
3. **Generic fallback** — if a type has no template but the row has `Fields["arn"]`, use AWS's ARN resolver: `https://<partition-console-domain>/go/view?arn=<url-encoded-arn>` (live-verified 2026-07-20: 302s to the right service page, region taken from the ARN, works on all three partition domains). If neither → flash "no console link for this type".
4. **Key wiring** — `Open`/`CopyURL` bindings in `internal/tui/keys/keys.go`; matched below the existing input-mode guards in `internal/tui/app_input.go` (the guards at :26-50 already protect filter/command/search/related-filter), gated by screen kind like `ToggleRelated` (:189); side effect in a `tea.Cmd` closure like `copyToClipboard` (`internal/tui/runtime_adapter_navigate.go:890-900`). Core's controller treats it like `ActionCopy` (`core/app/controller.go:467-471`): the *side effect* is renderer-only.
5. **Browser opener** — honors the `$BROWSER` environment variable first (standard convention), then falls back to per-GOOS exec (darwin `open`, linux `xdg-open`, windows `rundll32 url.dll,FileProtocolHandler`), argv form, no shell. No new dependency — matches how `atotto/clipboard` does per-OS work. This would be the first `exec.Command` in the product; it spawns a local browser only, no AWS API call, so the read-only principle holds. `$BROWSER` is also the seam the smoke tests use to capture opened URLs without launching a real browser.
6. **Web mode parity** — the SSE channel only signals "re-render"; the server cannot (and must not) exec a browser server-side. Instead the controller exposes the selected row's console URL in the snapshot, and `core/web/static/app.js` handles `o` locally with `window.open(url)` **inside the keydown handler** (popup blockers kill `window.open` from async callbacks). Bind bare `o` only — `Ctrl+O` is the browser's own file dialog and app.js deliberately lets ctrl-combos through.

## Verified console URL reference (implementation seed)

All patterns below were live-verified against AWS's own `/go/view` 302 redirects (not from memory). Canonical regional shape uses **both** the region subdomain and the `?region=` param — that is what AWS's own resolver emits.

| Type | URL pattern |
|---|---|
| EC2 instance | `https://{region}.console.aws.amazon.com/ec2/home?region={region}#InstanceDetails:instanceId={id}` |
| S3 bucket | `https://console.aws.amazon.com/s3/buckets/{name}` (regionless host, `?region=` optional) |
| RDS instance | `…/rds/home?region={region}#database:id={id};is-cluster=false` |
| RDS/Aurora cluster | `…/rds/home?region={region}#database:id={id};is-cluster=true` |
| DocumentDB cluster | `…/docdb/home?region={region}#cluster-details/{id}` (rds: ARN → wrong console via `/go/view`, must special-case by engine) |
| ElastiCache | `…/elasticache/home?region={region}#/redis/{id}` vs `#/memcached/{id}` vs `#/valkey/{id}` — branch on engine |
| EKS cluster | `…/eks/home?region={region}#/clusters/{name}` |
| EKS node group | `…/eks/home?region={region}#/clusters/{cluster}/nodegroups/{name}` |
| Secrets Manager | `…/secretsmanager/secret?region={region}&name={name}` (name = ARN resource minus random `-XXXXXX` suffix, URL-encoded) |
| VPC / SG / Subnet | `…/vpc/home?region={region}#VpcDetails:VpcId=…` / `#SecurityGroup:groupId=…` / `#SubnetDetails:subnetId=…` |
| Lambda | `…/lambda/home?region={region}#/functions/{name}` |
| CloudWatch alarm | `…/cloudwatch/home?region={region}#alarmsV2:alarm/{name}` (encode) |
| Log group | `…/cloudwatch/home?region={region}#logsV2:log-groups/log-group/{name with / → %2F}` |
| KMS key | `…/kms/home?region={region}#/kms/keys/{keyId}` |
| IAM role | `https://console.aws.amazon.com/iam/home#/roles/details/{name}` (global, no region) |
| ELB/ALB/NLB | `…/ec2/home?region={region}#LoadBalancer:loadBalancerArn={arn}` |
| ECS cluster | `…/ecs/v2/clusters/{name}?region={region}` |
| DynamoDB table | `…/dynamodbv2/home?region={region}#table?name={name}` |
| SQS queue | `…/sqs/v3/home?region={region}#/queues/{arn}` |
| SNS topic | `…/sns/v3/home?region={region}#/topic/{arn}` |
| CloudFormation stack | `…/cloudformation/home?region={region}#/stacks/stackinfo?stackId={stackArn}` |
| CloudTrail trail | `…/cloudtrailv2/home?region={region}#/trails/{arn}` |
| Route 53 zone | `https://console.aws.amazon.com/route53/v2/hostedzones#ListRecordSets/{zoneId}` (global) |
| CloudFront distribution | `https://console.aws.amazon.com/cloudfront/v4/home#/distributions/{id}` (global) |
| Generic fallback | `https://{partition-domain}/go/view?arn={url-encoded-arn}` — 302s to the right page; hard 404 for unsupported types |

Reference corpora for the remaining ~45 top-level types: `link2aws/link2aws.github.io` (per-type templates incl. partitions) and AWS-org `cdklabs/cdk-monitoring-constructs` `AwsConsoleUrlFactory.ts`.

## Edge cases and risks

**Identity & data model**

1. **No universal ARN.** `domain.Resource` has no ARN field; `Fields["arn"]` is a convention populated in Wave 1 for ~40 types (rds, secrets, lambda, ddb, sqs, …) but **not** for the whole EC2 family (ec2, sg, vpc, subnet, eni, ebs, ami, …) or s3. This is why the per-type template is primary and `/go/view?arn=` is only a fallback — most console URLs are region+ID shaped and don't need the ARN at all.
2. **Account ID is async and Rotate-cleared.** `Session.Identity` may be nil early in a session and after profile/region rotation. Templates must not require `{account}` except where truly needed; when needed and nil → flash instead of building a broken URL.
3. **Region comes from the session pair**, not the row. `Rotate()` clears row state anyway, so a row can never outlive its region — but the builder must read region at press time via the accessor, not cache it.

**Service quirks (why a plain template string isn't enough everywhere)**

4. **Global services**: IAM (role/policy/iam-user/iam-group), Route 53, CloudFront take no region subdomain/param; S3 bucket pages use the regionless host.
5. **WAFv2 scope**: `CLOUDFRONT` scope lives on the global console (us-east-1), `REGIONAL` on the region's console — branch on scope.
6. **Shared-console engines**: `dbc` is Aurora *or* DocumentDB (rds vs docdb console — `/go/view` gets this wrong); `redis` is redis/valkey/memcached/serverless (different hash routes). Branch on engine fields.
7. **Rows without a canonical page**: `ct-events` (CloudTrail events) has no per-event URL — best effort is the Event history search page, or flash "no console page". Same decision needed per child type (29 child types; log streams and Step Functions executions *do* have URLs — good wave-2 candidates).

**Environment & partition**

8. **Partitions**: China → `console.amazonaws.cn`, GovCloud → `console.amazonaws-us-gov.com`. Derive from the ARN partition field or region prefix (`cn-`, `us-gov-`). Verified gotcha: `/go/view` does **not** route cross-partition — the commercial endpoint happily "resolves" a GovCloud ARN onto the wrong domain, so the builder must pick the domain itself.
9. **Browser logged into the wrong account.** The URL is correct; the user's console session may be a different account/role → "not found" or, worse, a *same-named resource in the wrong account*. Can't be fixed client-side; document it in help. Future enhancement: IAM Identity Center **shortcut links** (`https://{sub}.awsapps.com/start/#/console?account_id=…&role_name=…&destination={encoded URL}` — officially documented) behind an optional config key, which pins account+role explicitly.
10. **Headless/SSH/no GUI**: `xdg-open` missing, `DISPLAY` unset, remote tmux — the exec fails; must flash the error, never crash or block (this is exactly what `O` copy-URL is for).
11. **Demo mode**: profile/region/account are synthetic; opening a real console on a fixture resource is misleading. Detect via the existing `IsDemo` flag (renderer and core both have it) and no-op with a flash — consistent with how demo mode already skips Wave-2 probes.

**Correctness & security**

12. **URL encoding is a minefield**: secret names contain `/`, log group names need `/`→`%2F` (console itself uses double-encoded `$252F` — both accepted), alarm names allow spaces/unicode, SQS/SNS fragments embed full ARNs. Every placeholder must go through proper query/path escaping — no string concatenation of raw names.
13. **No shell, ever**: resource names are untrusted input from AWS. The opener must use argv-form `exec.Command` (no `sh -c`), and the builder should assert the final URL is `https` on an allow-listed console domain before exec — a defense-in-depth guard that makes a crafted name harmless by construction.
14. **Console route churn**: the hash-route URL shapes are reverse-engineered and drift with console redesigns (sqs v2→v3, dynamodb→dynamodbv2, cloudtrail→cloudtrailv2, cloudfront v3→v4). Mitigations: old routes 302-redirect for years; the `/go/view` fallback is maintained server-side by AWS; all templates live in one package with per-type unit tests so a route bump is a one-line fix.
15. **Stale rows**: cached/deleted resources open a console 404 — acceptable and harmless (read-only).
16. **Help-text clarity**: `o` sits next to sort key `0` (zero) on help surfaces — label unambiguously ("o open in console").

## Scope

**v1**: `o` + `O` on top-level resource lists (all 70 types — template or documented flash), detail view, related-panel selection; demo-mode gating; TUI only-or-with-web-parity (web parity is cheap: snapshot field + keydown handler).
**Follow-ups (separate issues)**: child-view rows with real URLs (log streams, executions, builds, ECS tasks); Identity Center shortcut-link config; costs screen → Cost Explorer deep link; `:open` command alias.

## Acceptance criteria

- [ ] `o` on every top-level list opens the correct, correctly-encoded URL for the highlighted row — every one of the 70 types either has a verified mapping or deliberately flashes "no console page" (decision recorded per type; no silent fallthrough)
- [ ] `o` in detail view opens the detailed resource; with related panel focused, opens the selected related row's page
- [ ] `O` copies the same URL via the existing clipboard path with a flash
- [ ] Global services, WAF scope, dbc/redis engine branching, and partition domains (aws / aws-cn / aws-us-gov) produce correct URLs (unit-tested)
- [ ] Encoding: names with `/`, spaces, and unicode survive round-trip (unit-tested with hostile fixture names)
- [ ] Demo mode: no browser launch, flash shown; `--demo` smoke passes
- [ ] Opener failure (no `xdg-open`/`DISPLAY`) flashes an error, UI stays responsive
- [ ] No AWS API calls added; no shell interpolation; https + console-domain allow-list asserted before exec
- [ ] Web mode: `o` opens the page via `window.open` in the keydown handler; `Ctrl+O` untouched
- [ ] Input modes (`/`, `:`, search, related filter) still receive the literal character
- [ ] Smoke coverage: demo smoke (pre-push gate) asserts the demo flash + no browser spawn for `o`, clipboard-verified URL correctness for `O` across a representative type sample, and input-mode passthrough; the live traverse gains the `$BROWSER` capture step for `o` on real rows
- [ ] Docs updated: `docs/shared/keybindings.md` + README regenerated via readmegen, `core/domain/helpkeys.go` sections, footer hints if added, `app.js` keyMap, GitHub wiki Key Bindings page at release

## Test plan (TDD)

- `a9s-qa` first: table-driven unit tests for the URL builder covering **all** top-level types using the shared demo fixtures (region/partition/encoding/engine matrix); key-wiring tests (list, detail, related-panel, demo gating, input-mode passthrough) via the scenario harness — no interactive runs.
- `a9s-coder`: builder package, catalog field + per-type entries, key wiring, opener, web parity.
- Opener itself: injected as an interface so unit tests assert the URL handed to it without spawning processes.

### Real smoke tests (tmux, real binary, rendered-surface assertions)

Unit tests prove the builder in isolation; the smokes prove the whole path through the real binary — key dispatch, session region, catalog lookup, encoding, flash rendering. Three tiers:

1. **Demo smoke — extends the `a9s-smoke` BASE walk, runs in the `make ready-to-push` gate.** Against `./a9s --demo` in tmux:
   - navigate to a list, press `o` → assert the "demo mode" flash renders on the surface and **no browser process spawned** (set `$BROWSER` to a sentinel script that writes a file; assert the file does not appear);
   - press `O` on rows of several fixture types (regional, global, engine-branched — e.g. ec2, s3, role, dbc/DocumentDB, secrets, logs) → assert the copy flash, then read the clipboard (`pbpaste`) and assert a well-formed `https` console URL: correct partition domain, demo-session region pinned in both subdomain and `?region=`, ID/name correctly encoded. This is a full end-to-end URL check through the real binary with **zero AWS calls and zero browser launches** — it works precisely because `O` copy stays enabled in demo mode;
   - press `o` while in filter `/` and command `:` modes → assert the character lands in the input, no flash, nothing opened.
2. **`$BROWSER` capture smoke — part of the `a9s-traverse` live walk (readonly account), not the pre-push gate.** Set `BROWSER` to a capture script that appends each URL to a file; walk a representative sample of types (per-type checklist), press `o` on a real row, assert each captured URL is `https`, on the allow-listed console domain, and carries the live session's region and the row's real ID. No real browser opens during the traverse.
3. **Route-liveness probe (same traverse, optional per run):** `curl -sI` each captured URL and assert the console answers with a redirect (sign-in or resource page), **not** a 404 — this is the tripwire for console route churn (edge case 14) that no offline test can catch.

The smoke scenarios follow `tests/integration/SCENARIO_HARNESS.md` — never a hand-driven interactive session.

## Non-goals

- Any AWS write operation (unchanged: a9s never calls mutating APIs)
- Embedded browser / auth handling — a9s opens the URL; the console session is the user's problem
- Per-user key rebinding (no runtime rebinding exists today)

### PR DESCRIPTION
**Pull request by k2m30**
Closes #477.

`o` opens the highlighted resource's exact AWS console page; `O` copies the URL. Coverage: all 70 top-level and 29 child-view types, on lists, child lists, detail, and the focused related-panel row; web mode has full parity (client-side `window.open`/clipboard — the server never execs).

Highlights:

- Every URL shape live-verified against AWS's own `/go/view` ARN resolver (78 captures incl. GovCloud/China domains) + official docs; per-type config, no heuristics; builders return no-link instead of a wrong-but-plausible URL on missing input
- One controller-owned resolver (`ConsoleTarget`) feeds both renderers; new `core/consolelink` package with an https + console-domain allow-list guard before any exec
- `$BROWSER` honored (argv-split, never a shell), per-OS fallback, child reaped
- Demo mode flashes instead of opening; `O` still copies
- Docs: keybindings + architecture aligned; new wiki Environment Variables page (README/website point there); CHANGELOG + `releases/v3.56.0.md` for the v3.56.0 minor

Review trail so far: two Codex rounds (5 findings, each fixed + regression-tested), a full self-review (2 hardening guards), tmux smoke of the compiled binary, real-browser web verification. `make ready-to-push` green; `make ready-to-release` (live readonly smokes) running as part of this release.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: core/aws/catalog_messaging.go [hunk 1]
```diff
@@ -208,6 +242,13 @@ var messagingTypes = []catalog.ResourceTypeDef{ //nolint:gochecknoglobals // sta
 		Aliases:       []string{"sqs", "queues"},
 		Category:      "MESSAGING",
 		CloudTrailKey: "ResourceName:Fields.arn",
+		ConsoleURL: func(r domain.Resource, region, _ string) string {
+			arn := r.Fields["arn"]
+			if arn == "" {
+				return ""
+			}
+			return consolelink.Regional(region, "sqs/v3/home?region="+region+"#/queues/"+arn)
+		},
```
[the remark this answers is not in the document]
Comment (k2m30) [reply core/aws/catalog_messaging.go:251]: Disproved by live capture from AWS's own console resolver (GET https://console.aws.amazon.com/go/view?arn=arn:aws:sqs:eu-central-1:123456789012:my-queue, 2026-07-20): the 302 Location is https://eu-central-1.console.aws.amazon.com/sqs/v3/home?region=eu-central-1#/queues/arn:aws:sqs:eu-central-1:123456789012:my-queue — the raw ARN in the fragment, exactly the shape this builder emits. The URL-encoded queue-URL form also works, but the ARN form is what AWS itself currently produces, and the per-type test pins it.

File: core/aws/catalog_monitoring.go [hunk 2]
```diff
@@ -131,6 +136,9 @@ var monitoringTypes = []catalog.ResourceTypeDef{ //nolint:gochecknoglobals // st
 		Aliases:       []string{"logs", "loggroups", "log-groups", "cwlogs", "log_groups"},
 		Category:      "MONITORING",
 		CloudTrailKey: "ResourceName:ID",
+		ConsoleURL: func(r domain.Resource, region, _ string) string {
+			return consolelink.Regional(region, "cloudwatch/home?region="+region+"#logsV2:log-groups/log-group/"+url.PathEscape(r.ID))
+		},
```
[the remark this answers is not in the document]
Comment (k2m30) [reply core/aws/catalog_monitoring.go:141]: Disproved by live capture from AWS's own console resolver (2026-07-20): for arn:aws:logs:...:log-group:/aws/lambda/my-fn the 302 Location is ...cloudwatch/home?region=eu-central-1#logsV2:log-groups/log-group/%2Faws%2Flambda%2Fmy-fn — single-encoded %2F, which is exactly what url.PathEscape produces here. The log-stream capture (…/log-events/2026%2F07%2F20%2Fmystream) confirms the same for the shared helper. The $252F double-encoded form seen in console-copied links also works, but %2F is AWS's own current emission and does not break the router.

File: core/aws/catalog_secrets.go [hunk 3]
```diff
@@ -130,6 +135,10 @@ var secretsTypes = []catalog.ResourceTypeDef{ //nolint:gochecknoglobals // stati
 		Aliases:       []string{"ssm", "parameters", "parameter-store"},
 		Category:      "SECRETS & CONFIG",
 		CloudTrailKey: "ResourceName:ID",
+		ConsoleURL: func(r domain.Resource, region, _ string) string {
+			name := strings.TrimPrefix(r.ID, "/")
+			return consolelink.Regional(region, "systems-manager/parameters/"+name+"/description?region="+region)
+		},
```
[the remark this answers is not in the document]
Comment (k2m30) [reply core/aws/catalog_secrets.go:141]: Disproved by live capture from AWS's own console resolver (2026-07-20): for arn:aws:ssm:...:parameter/my/param the 302 Location is ...console.aws.amazon.com/systems-manager/parameters/my/param/description?region=eu-central-1 — raw slashes in the path, exactly what this builder emits. PathEscape would turn them into %2F and deviate from the verified shape; the hostile-input unit test (TestConsoleURL_HostileInput_SSMParameterNameEncoding) pins the raw-slash contract.

File: core/aws/catalog_security.go [hunk 4]
```diff
@@ -285,6 +303,17 @@ var securityTypes = []catalog.ResourceTypeDef{ //nolint:gochecknoglobals // stat
...
+			scope := r.Fields["scope"]
+			if r.Name == "" || scope == "" {
+				return ""
+			}
+			tail := "wafv2-pro/protections/" + url.PathEscape(r.Name) + "/" + r.ID + "?panel=protectionPackHome"
+			if scope == "CLOUDFRONT" {
+				return consolelink.Regional("us-east-1", tail+"&region=us-east-1&scope=global")
+			}
+			return consolelink.Regional(region, tail+"&region="+region+"&scope=regional")
+		},
```
[the remark this answers is not in the document]
Comment (k2m30) [reply core/aws/catalog_security.go:316]: Disproved: domain.Resource has an exported Name field (core/domain/resource.go) — this compiles (the full gate is green on this tree) and the waf fetcher populates Name with the Web ACL name for every Wave-1 row, so the deep link renders correctly (pinned by TestConsoleURL_PerType/waf against the shared fixtures, plus the dedicated REGIONAL/CLOUDFRONT branch tests). The r.Name == "" guard exists precisely for field-less stub rows from the related panel, where the builder must return no-link (TestConsoleURL_IncompleteRow_Waf_EmptyNameOrScopeReturnsEmpty).
